### DISCUSSION TITLE
Failed to fetch oauth token from Ring

### DISCUSSION THREAD

**Discussion by mlohus93** [tsightler/ring-mqtt#342] [Support Q&A] [answered]
I am grateful for the skill and expertise provided by the author of MQTT for Ring, and those who would monitor these Q&A boards. 
 I hope this isnt a silly error, I have tried researching every way I can think of but cannot seem to resole. 

I have used the standard Ring Integration for Home Assistant for the past 12+ months quite successfully, I started having problems sometime since the first of the year with the integration not letting me log in or do 2FA.  I read that someone resolved by switching to Ring MQTT with Video Streaming - an Addon for Home Assistant.  I tried that but I still have the same problem.  Rebooting has not helped, waiting has not helped.  If I try to login, using credentials I have just tested successfully on the Ring Website, I keep getting this message and it never triggers a 2FA message to my phone and never displays the 2FA screen:

- Acquire Refresh Token
- Please enter your Ring account information below and the 2FA code on the following page to generate the refresh token required for this application to access the Ring API.
- Failed to fetch oauth token from Ring. Verify that your email and password are correct. (error: access_denied)

I am new to MQTT, but I believe I have my MQTT and MQTT Broker installed and configured correctly - Log is below... what am I doing wrong?  Appreciate any assistance.

[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] ring-mqtt.sh: executing... 
-------------------------------------------------------
| Ring-MQTT with Video Streaming                      |
| Addon for Home Assistant                            |
|                                                     |
| Report issues at:                                   |
| https://github.com/tsightler/ring-mqtt-hassio-addon |
-------------------------------------------------------
[cont-init.d] ring-mqtt.sh: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
-------------------------------------------------------
ring-mqtt.js version: 4.9.1
Node version v16.13.2
NPM version 8.1.3
git version 2.34.1
-------------------------------------------------------
Using discovered MQTT Host: core-mosquitto
Using discovered MQTT Port: 1883
Using configured MQTT User: mqtt-buzz
Using configured MQTT password: <hidden>
-------------------------------------------------------
Running ring-mqtt...
2022-03-26T20:31:29.665Z ring-mqtt Using configuration file: /data/options.json
2022-03-26T20:31:29.670Z ring-mqtt File /data/ring-state.json not found. No saved state data available.
2022-03-26T20:31:29.671Z ring-mqtt No refresh token was found in saved state file or config file.
2022-03-26T20:31:29.671Z ring-mqtt Use the web interface to generate a new token.
 
Here is my Ring-MQTT with Video Streaming configuration:

enable_cameras: true
snapshot_mode: enabled
livestream_user: ''
livestream_pass: ''
enable_modes: false
enable_panic: false
beam_duration: 0
disarm_code: ''
mqtt_host: <auto_detect>
mqtt_port: <auto_detect>
mqtt_user: mqtt-buzz
mqtt_password: PASSWORD-REDACTED
branch: addon
debug: ring-*
ring_token: ''
location_ids:
  - ''

**Comment by tsightler** [comment, marked as answer]
Sorry, I didn't see this message previously.  I don't have any significant suggestions for you as authentication is very basic and works for 1000's of ring-mqtt users.  Is it possible you have some special character in your username or password that might not be trapped and encoded correctly?  That's really the only thing I can think of.  The fact that you are having the issue with both the built-in integration and this one seems to indicate something unique to your account.  You could try creating a dedicate account just for the addon using a different username and just share the required devices to that account.  This has a few limitations (for example chimes can't be shared), but 95% of features will work this way and it might be a good test.

Personally, I'd also suggest changing the password on the account even though you say it works, it could be some kind of weird issue with a stale token.  Good luck!

**Reply by mlohus93** [reply]
Thank you so much. Honestly it never occurred to me that my user-id (email) could be a problem.  I changed from an email with a period in the middle (xxxxxx.yyyyyy@zzz.com) to a different email I already had without the period (xxxxxx@zzz.com) and the 2FA request worked in both MQTT and native integrations first try. Either it was that period or it was the length prior to the @ symbol as the new one is shorter.   Grateful to have that resolved.

**Reply by tsightler** [reply]
Thanks for coming back and I'm glad you made progress.  I don't think the period should be a problem as that's pretty common and I'd think there would be other complaints about that issue, but I'll give it a test when I can, perhaps somehow it's not being encoded.
